### PR TITLE
NanoId faster and shorter than uuid

### LINKED ISSUE #16912: NanoId faster and shorter than uuid
Nonaid which is a distributed ID generation method, faster and shorter than uuid

<html>
<body>
<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

Test case | nanoid（ms） | uuid（ms）
-- | -- | --
1 thread generate total 10000000 | 15252 | 12546
1 thread generate total 10000000，and ThreadLocalRandom optimization | 3596 | 5824
10 thread generate total 10000000 | 23909 | 15609
10 thread generate total 10000000，and ThreadLocalRandom optimization | 4094 | 5521
test environment：Linux version 3.10.0-1062.18.1.el7.x86_64，vm 8c 16m  conclusion：Nanoid performs better than UUID when optimized using ThreadLocalRandom

</div>
</body>
</html>

NanoIdUtils.randomNanoId(ThreadLocalRandom.current(), NanoIdUtils.DEFAULT_ALPHABET, NanoIdUtils.DEFAULT_SIZE);
new UUID(ThreadLocalRandom.current().nextLong(), ThreadLocalRandom.current().nextLong()).toString();

### PR DESCRIPTION
**Pull request by lwclover**
Fixes #16912.

### CODE REVIEW

**Review by tuichenchuxin — DISMISSED** [review]

**Review by terrymanu — CHANGES_REQUESTED** [review]
Please check the licenses of the new dependencies and update the license files

**Review by terrymanu — CHANGES_REQUESTED** [review]
The license is MIT which is fine, please add it into license file

**Review by TeslaCN — COMMENTED** [review]
Maybe the nanoid is shorter. But is there any benchmark to prove it faster than UUID? The default UUID constructor uses the SecureRandom which may be slow under high concurrency, but we may use the `ThreadLocalRandom` when generating UUID, just like what we did in ExecutionGroupContext.

**Comment by TeslaCN** [comment]
It it better to compare UUID generated by ThreadLocalRandom with NanoId.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/NonaIdKeyGenerateAlgorithm.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+package org.apache.shardingsphere.sharding.algorithm.keygen;
+
+import org.apache.shardingsphere.sharding.spi.KeyGenerateAlgorithm;
+
+import com.aventrix.jnanoid.jnanoid.NanoIdUtils;
+
+/**
+ * NonaId key generate algorithm.
+ */
+public final class NonaIdKeyGenerateAlgorithm implements KeyGenerateAlgorithm {
```
Comment (TeslaCN) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/NonaIdKeyGenerateAlgorithm.java:27]: `Nona` or `Nano`?

**Review by TeslaCN — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/NonaIdKeyGenerateAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    }
+    
+    @Override
+    public Comparable<?> generateKey() {
+        return NanoIdUtils.randomNanoId();
+    }
+    
+    @Override
+    public String getType() {
+        return "NONAID";
```
Comment (TeslaCN) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/NonaIdKeyGenerateAlgorithm.java:40]: Same as above.

**Review by TeslaCN — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/NonaIdKeyGenerateAlgorithm.java (hunk 1 above)
Comment (lwclover) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/NonaIdKeyGenerateAlgorithm.java]: done

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/NonaIdKeyGenerateAlgorithm.java (hunk 2 above)
Comment (lwclover) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/keygen/NonaIdKeyGenerateAlgorithm.java]: done

**Comment by lwclover** [comment]
> It it better to perform a JMH test to compare UUID generated by ThreadLocalRandom with NanoId.

<html>
<body>
<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div data-zone-id="0" data-line-index="0" style="white-space: pre;">

Test case | nanoid（ms） | uuid（ms）
-- | -- | --
1 thread generate total 10000000 | 15252 | 12546
1 thread generate total 10000000，and ThreadLocalRandom optimization | 3596 | 5824
10 thread generate total 10000000 | 23909 | 15609
10 thread generate total 10000000，and ThreadLocalRandom optimization | 4094 | 5521
test environment：Linux version 3.10.0-1062.18.1.el7.x86_64，8c 16m  conclusion：Nanoid performs better than UUID when optimized using ThreadLocalRandom

</div>
</body>
</html>

**Review by terrymanu — APPROVED** [review]

**Review by TeslaCN — APPROVED** [review]

**Review by tuichenchuxin — APPROVED** [review]

**Pull request merged**